### ISSUE TITLE
[Infra] 실험 재현성·공정 비교를 코드가 보장하지 않음 — split·시드 일치 자동검증·스냅샷 provenance 부재

### ISSUE THREAD

**Issue by bbungjun** [SKYAHO/Autoresearch#423]
## 배경

Capability probe(agent 서버 도입 전, 로컬 에이전트에게 "가설→피처준비→모델구성→학습평가→정책시뮬레이션→승격대기" 6단계를 실제로 맡겨 어디가 막히는지 실측하는 실험) round_001(FM vs LightGBM), round_002(MLP vs LightGBM) 두 라운드에서 공통으로 드러난 문제입니다.

> **책임 분리**: 필수 환경변수 fail-fast 가드(BigQuery 접속 전 `GCS_REGISTRY_PATH`/`GCS_STAGING_LOCATION` 확인)는 #404 완료 조건에 이미 포함되어 있어 이 이슈에서는 제외합니다. 경위는 코멘트 참고.

## 문제 1 — 같은 스냅샷·split·시드로 학습했는지 자동 검증이 없음

모델 비교(baseline vs challenger) 실험에서 "공정 비교" 전제(같은 데이터 스냅샷, 같은 train/test split, 같은 시드)를 코드가 전혀 보증해주지 않습니다. 두 라운드 모두 held-out test set이 실제로 같은 행인지 `md5sum`으로 직접 대조하는 절차를 매번 즉흥적으로 만들어 확인해야 했습니다(저장소 내 `md5` 관련 코드/테스트 0건). 격차가 크면(round_001, −25pp) 문제 없지만, 격차가 작으면(round_002, +2.63pp) 이 검증 없이는 결과를 신뢰할 수 없습니다.

## 문제 2 — 데이터 스냅샷에 provenance 메타데이터가 없음

학습에 쓰는 CSV 스냅샷에 조회 기간·생성 시각·FeatureService 버전 같은 이력 정보가 전혀 없습니다. 스냅샷 파일 자체가 순수 CSV라 "이게 언제, 어떤 조건으로 만들어졌는지" 추적이 불가능합니다.

## 왜 중요한가

자율 실험 에이전트가 여러 라운드를 반복 수행하려면 각 실험이 재현 가능해야 하는데, 지금은 재현성·공정성 검증 자체가 사람(에이전트)의 즉흥적 판단에 의존합니다.

## 제안 범위

- baseline/challenger 실행 메타데이터(스냅샷 해시, split 파라미터, seed)를 자동 비교하는 표준 검증 스텝
- 데이터 스냅샷 매니페스트(조회기간·생성시각) 파일 규약

## 관련

- #404 — 필수 환경변수 fail-fast 가드(이 이슈에서 제외한 부분)를 담당. 책임 분리 경위는 코멘트 참고.

**Comment by waieiches** [comment]
## 겹치는 부분 대조 (전체는 #403에)

### 문제 1 — #404 완료조건 2와 원하는 건 같지만, 고칠 자리가 다릅니다

#404 완료조건 2가 "자격 증명·feast 준비가 스킬 안에서 처리되거나, 없을 때 명확한 안내와 함께 중단한다"입니다. 여기까진 이 이슈와 같습니다.

다른 건 어디를 고치느냐입니다.

- **#404** → 새로 만들 스킬(래퍼) 안에서 중단
- **이 이슈** → 지금 있는 `_assemble_via_feast`의 실행 순서

서로 대신하지 못합니다. 스킬만 만들면 CLI를 직접 부를 때 여전히 매달리고, 순서만 고치면 스킬은 여전히 없습니다. **둘 다 필요합니다.**

**한 가지 짚어드릴 게 있습니다.** 지목하신 `load_training_entity_spine()`은 사실 그 환경변수를 읽지 않습니다. 읽는 건 부르는 쪽입니다.

```python
def _assemble_via_feast(...):
    spine = load_training_entity_spine(...)           # BigQuery 접속 (여기서 매달림)
    ...
    registry_path = os.environ["GCS_REGISTRY_PATH"]   # "설정 없음" 오류가 나야 할 자리
    gcs_staging  = os.environ["GCS_STAGING_LOCATION"]
```

증상은 정확히 보신 게 맞고, **고칠 자리가 호출하는 쪽**이라는 차이입니다. 아래 두 줄을 위로 올리면 되는 크기로 보입니다.

### 문제 2 — ⑥과는 다른 요구입니다

- **⑥(champion 공정 비교)** : champion이 **어떤 조건으로 학습됐는지 알아내서 똑같이 재현**하기
- **문제 2** : 두 실험이 같은 조건이었는지 **끝나고 확인**하기

앞은 재현, 뒤는 확인입니다. 문제 2는 champion과 상관없이 baseline vs challenger 비교에서도 필요하니 **따로 닫힙니다.**

다만 구현을 공유할 수는 있어 보입니다(실행 메타데이터를 비교하는 유틸). ⑥은 지금 따로 이슈 없이 #404 완료 조건에 들어가 있는데, **이 이슈에서 함께 다루는 게 나을지 그대로 두는 게 나을지 정해 주시면 따르겠습니다.**

### 문제 3 — 절반은 이미 있습니다

이력이 "전혀 없다"기보다 **어디에 남는지가 갈려 있습니다.** #359 C2 작업으로 MLflow run에는 이미 남고 있습니다 (`src/cli.py`의 `data_source_params`).

```
assembly_source, feature_service, events_start_date, events_end_date, feast_registry_path
```

없는 건 **CSV 파일 자체에 붙는** 정보입니다. 지금 상태는 "MLflow를 거쳐 보면 알 수 있는데, CSV만 들고 있으면 알 수 없다"입니다. 이 기준으로 범위를 좁히면 할 일이 줄 것 같습니다.

---

문제1은 두 줄 옮기는 크기라 먼저 고쳐두면 오히려 #404가 편해집니다. 지금 진행하셔도 될지 의견 주시면 좋겠습니다.

**Comment by bbungjun** [comment]
## 책임 분리 정리

#404와 범위가 겹쳐 있어 정리합니다.

- **필수 환경변수 fail-fast 가드**(`load_training_entity_spine()`이 BigQuery 접속 전에 `GCS_REGISTRY_PATH`/`GCS_STAGING_LOCATION` 등을 먼저 확인해야 한다는 문제, 기존 "문제 1")는 **#404의 완료 조건에 이미 포함**되어 있습니다("자격 증명·feast 환경 준비가 스킬 안에서 처리되거나, 없을 때 명확한 안내와 함께 중단한다"). 중복 발행이라 이 이슈에서 제거합니다.
- 이 이슈는 **split/seed 일치 자동검증**과 **스냅샷 provenance 메타데이터** 두 가지만 남깁니다. 둘 다 #404(단일 조립 스킬)와는 다른 산출물(비교 검증 스텝, 매니페스트 규약)이라 겹치지 않습니다.

제목·본문을 좁혀서 다시 작성합니다.
